ch_pipeline/processing/base.py

```diff
@@ -347,26 +358,71 @@ def generate(self, max=10, submit=True):
+        to_run = self._generate_hook(user=user)[:max]
 
         for tag in to_run:
             queue_job(self.job_script(tag), submit=submit)
 
-    def _generate_hook(self):
+    def _generate_hook(self, user=None):
         """Override to add custom behaviour when jobs are queued."""
-        return self.pending()
+        return self.status(user=user)["not_yet_submitted"]
 
     def pending(self, user=None):
         """Jobs available to run."""
+        warnings.warn(
+            "'pending' method is deprecated. Call 'status()['not_yet_submitted']'."
+        )
+        return self.status(user)["not_yet_submitted"]
 
-        waiting, running = self.queued(user)
-        not_pending = set(self.ls()) | set(waiting) | set(running)
-        pending = [job for job in self.available() if job not in not_pending]
+    def failed(self, user=None, time_sort=False) -> Dict[str, list]:
+        """Categorize failed jobs.
+
+        Parameters
+        ----------
+        user : str
+            user to find running jobs for. If not provided, this
+            will default to the current user
 
-        return pending
+        Returns
+        -------
+        crashed : dict
+            tags associated with each category
+        """
+        # Get a set of all tags to check for cause of failure
+        failed_tags = self.status(user, time_sort)["failed"]
+        # Regex patterns associated with common sources of failure
+        # Keys can be inserted in order of priority/specificity
+        # as of Python 3.7+, as dicts will preserve insertion order.
+        # This could be buggy on versions of python below 3.7, as the
+        # broad 'unknown' key could match before a more specfic one
+        patterns = {
+            # Find crashes due to unwanted NaN/Inf values
+            "nan_or_inf": [
+                re.compile(r"ValueError(.*?)infs(.*?)NaNs"),
+            ],
+            # Find out of memory errors
+            "out_of_memory": [
+                re.compile(r"numpy.core._exceptions._ArrayMemoryError(.*?)MPI_ABORT")
+            ],
+            # Error connecting to chimedb database
+            "chimedb_error": [
+                re.compile(r"chimedb.core.exceptions.ConnectionError(.*?)MPI_ABORT")
+            ],
+            # Find crashes due to job timeout
+            "time_limit": [
+                re.compile(r"slurmstepd: error:(.*?)CANCELLED(.*?)TIME LIMIT"),
+            ],
+            "mpi_error": [
+                re.compile(r"draco.core.misc.CheckMPIEnvironment: MPI test failed")
+            ],
+            # Match anything else
+            "unknown": [re.compile(r"(.*?)")],
+        }
+        # Return all tags which match the listed patterns
+        return classify_failed(self.workdir_path, failed_tags, patterns)
 
-    def status(self, user=None) -> dict:
-        """Find the status of running jobs. This duplicates
+    def status(self, user=None, time_sort=False) -> dict:
```

Comment:
Again, docstring and type annotations.

Also be consistent with `Dict` vs `dict` (requires 3.9+) for annotations.